sdk/bulk.go

```diff
@@ -0,0 +1,145 @@
+package sdk
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/opentdf/platform/protocol/go/kas"
+)
+
+type BulkTDF struct {
+	Reader io.ReadSeeker
+	Writer io.Writer
+	Error  error
+}
+
+type BulkDecryptRequest struct {
+	TDFs    []*BulkTDF
+	TDFType TdfType
+}
+
+type BulkErrors []error
+
+func (b BulkErrors) Error() string {
+	return fmt.Sprintf("Some TDFs could not be Decrypted: %s", errors.Join(b...).Error())
+}
+
+// FromBulkErrors Returns List of Decrypt Failures and true if is decryption failures
+func FromBulkErrors(err error) ([]error, bool) {
+	var list BulkErrors
+	ok := errors.As(err, &list)
+	return list, ok
+}
+
+type BulkDecryptOption func(request *BulkDecryptRequest)
+
+func WithTDFs(tdfs ...*BulkTDF) BulkDecryptOption {
+	return func(request *BulkDecryptRequest) {
+		request.appendTDFs(tdfs...)
+	}
+}
+func WithTDFType(tdfType TdfType) BulkDecryptOption {
+	return func(request *BulkDecryptRequest) {
+		request.TDFType = tdfType
+	}
+}
+
+func createBulkRewrapRequest(options ...BulkDecryptOption) *BulkDecryptRequest {
+	req := &BulkDecryptRequest{}
+	for _, opt := range options {
+		opt(req)
+	}
+	return req
+}
+
+func (s SDK) createDecryptor(tdf *BulkTDF, tdfType TdfType) (decryptor, error) {
+	switch tdfType {
+	case Nano:
+		decryptor := createNanoTDFDecryptHandler(tdf.Reader, tdf.Writer)
+		return decryptor, nil
+	case Standard:
+		return s.createTDF3DecryptHandler(tdf.Writer, tdf.Reader)
+	case Invalid:
+	}
+	return nil, fmt.Errorf("unknown tdf type: %s", tdfType)
+}
+
+// BulkDecrypt
+func (s SDK) BulkDecrypt(ctx context.Context, opts ...BulkDecryptOption) error {
```

Comment:
There is a side of me that feels like the consumer of the sdk should decide when to actually decrypt payload. There might be a use case where someone would want to access encrypted metadata but not decrypt the full payload. Or maybe they want to do a random read to see if its actually the correct payload they need before actually decrypting it.

Reply:
i agree, although i think this feature should be seperate PR. NanoTDF lib would take some refactoring to get this feature working. Futhermore, I think as Bulk Go API will eventually need some more thoughtful features to be added into it, i.e. accessing manifests, collections, and encrypted metadata.  

Originally this is why I had `BulkDecryptRequest`.

Reply:
I agree with you both, I think we should have the option to defer the decrypt portion of the operation, but I think that can come in a future ticket.